stl/inc/syncstream

```diff
@@ -0,0 +1,312 @@
+// syncstream standard header
+
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#pragma once
+#ifndef _SYNCSTREAM_
+#define _SYNCSTREAM_
+#include <yvals.h>
+#if _STL_COMPILER_PREPROCESSOR
+
+#ifdef _M_CEE_PURE
+#error <syncstream> is not supported when compiling with /clr:pure
+#endif // _M_CEE_PURE
+
+#ifndef __cpp_lib_syncbuf
+#pragma message("The contents of <syncstream> are available only with C++20 syncbuf support.")
+#else // ^^^ !defined(__cpp_lib_syncbuf) / defined(__cpp_lib_syncbuf) vvv
+#include <map>
+#include <memory>
+#include <shared_mutex>
+#include <streambuf>
+
+#pragma pack(push, _CRT_PACKING)
+#pragma warning(push, _STL_WARNING_LEVEL)
+#pragma warning(disable : _STL_DISABLED_WARNINGS)
+_STL_DISABLE_CLANG_WARNINGS
+#pragma push_macro("new")
+#pragma push_macro("emit")
+#undef new
+#undef emit
+
+_STD_BEGIN
+
+struct _Mutex_count_pair {
+    shared_mutex _Mutex;
+    uint64_t _Ref_count = 0;
+};
+
+_CRTIMP2_PURE map<void*, _Mutex_count_pair>& __cdecl _Get_mutex_map();
+_CRTIMP2_PURE shared_mutex& __cdecl _Get_map_mutex();
+
+shared_mutex& _Get_mutex_for_instance(void* _Ptr) {
+    scoped_lock _Guard(_Get_map_mutex());
+    auto& _Mutex_map         = _Get_mutex_map();
+    auto _Instance_mutex_itr = _Mutex_map.find(_Ptr);
+    _STL_ASSERT(_Instance_mutex_itr != _Mutex_map.end(), "No mutex exists for given instance!");
+    return _Instance_mutex_itr->second._Mutex;
+}
+void _Add_mutex_for_instance_or_increment(void* _Ptr) noexcept {
+    scoped_lock _Guard(_Get_map_mutex());
+    auto& _Mutex_map = _Get_mutex_map();
+    _Mutex_map.try_emplace(_Ptr).first->second._Ref_count++;
+}
+void _Delete_mutex_for_instance_or_decrement(void* _Ptr) noexcept {
+    scoped_lock _Guard(_Get_map_mutex());
+    auto& _Mutex_map         = _Get_mutex_map();
+    auto _Instance_mutex_itr = _Mutex_map.find(_Ptr);
+    if (_Instance_mutex_itr == _Mutex_map.end())
+        return;
+    _Instance_mutex_itr->second._Ref_count--;

```

Comment:
We conventionally prefer predecrement by default; as `->` and `.` bind most tightly, no parentheses are needed.

I see that this somewhat verbose expression is then read on the next line; using predecrement would allow this to be combined with the test against `== 0`. That's probably a clarity win. (Expressions that do a *lot* of modifications are typically found in our oldest code - we prefer to avoid that nowadays, but a "decrement and test if zero" is pretty simple.)